File: src/main/golang/design/Stream.go
package design

import "reflect"

type Stream struct {
	value  interface{}
	filter func(interface{}) bool
	mapper func(interface{}) interface{}
}

func From(value ...interface{}) *Stream {
	return &Stream{value: value}
}

func (p *Stream) Where(filter func(interface{}) bool) *Stream {
	p.filter = filter
	return p
}

func (p *Stream) Map(mapper func(interface{}) interface{}) *Stream {
	p.mapper = mapper
	return p
}

func (p *Stream) Get() interface{} {
	val := reflect.ValueOf(p.value)

	if val.Kind() == reflect.Slice {
		var result []interface{}
		for i := 0; i < val.Len(); i++ {
			elem := val.Index(i).Interface()
			if p.filter == nil || p.filter(elem) {
				if p.mapper != nil {
					result = append(result, p.mapper(elem))
				} else {
					result = append(result, elem)
				}
			}
		}
		return result
	}

	if p.filter != nil && !p.filter(p.value) {
		return nil
	}
	if p.mapper != nil {
		return p.mapper(p.value)
	}
	return p.value
}

File: src/main/nodejs/Stream.ts
class Stream<T> {
    private readonly data: T[];
    private readonly pipeline: ((data: any[]) => any[])[];

    constructor(data: T[] | T, pipeline: ((data: any[]) => any[])[] = []) {
        this.data = Array.isArray(data) ? data : [data];
        this.pipeline = pipeline;
    }

    static from<T>(data: T[] | T): Stream<T> {
        return new Stream<T>(data);
    }

    where(predicate: (item: T) => boolean): Stream<T> {
        return new Stream<T>(this.data, [
            ...this.pipeline,
            (data: T[]) => data.filter(predicate),
        ]);
    }

    map<U>(mapper: (item: T) => U): Stream<U> {
        // @ts-ignore
        return new Stream<U>(this.data, [
            ...this.pipeline,
            (data: T[]) => data.map(mapper),
        ]);
    }

    first(): T | null {
        const result = this.execute();
        return result.length > 0 ? result[0] : null;
    }

    get(): T[] {
        return this.execute();
    }

    private execute(): T[] {
        return this.pipeline.reduce((acc, fn) => fn(acc), this.data);
    }
}
